### ISSUE TITLE
What are proving times like?

### ISSUE THREAD

**Issue by SilentCicero** [powdr-labs/powdr#399]
Great work on getting Powdr to this stage. It looks awesome. Also @emilyaherbert really well done picking a great team :)

I'm curious what proving times are like and if there is a toy Solidity verifier available?

Also, is there any work going into parallizable proving (I believe Risc0 has this)?

**Comment by leonardoalt** [comment]
We chatted in private about this but I'll answer here too so other people have access to the info:

The current prover we have fully integrated is Halo2 (PSE fork). It gives you direct access to a SNARK, but proving times are rather high. It is possible to generate a Yul verifier for Halo2 proofs, though the CLI doesn't do that yet. I just made https://github.com/powdr-labs/powdr/issues/415 to track that. The proofs can already be parallelized and aggregated, but that mechanism has to be built on top of the current crates.

We're working on modularizing the Polygon Hermez zkEVM prover (which we already support for the first STARK) to also be used that way (https://github.com/powdr-labs/zkevm-prover/issues/3).

On the VM side, we're adding support to Valida VM which also supports LLVM (next to RISCV). This + recursion will allow for some nice benchmarks.

tl;dr: we have lots of moving parts being built atm, full support and benchmarks soon™ .

**Comment by leonardoalt** [comment]
We're using https://github.com/powdr-labs/powdr-revme as benchmarks, though we haven't fully setup the Rust benches yet for performance. I'll close this specific issue for now, but happy to keep discussing in the other repo, DMs, or other issues.
